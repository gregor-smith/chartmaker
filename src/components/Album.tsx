import React, { FC } from 'react'
import { css, cx } from 'emotion'
import { Album as AlbumDetails, formatAlbumTitle } from '../state'


type Props = {
    details: AlbumDetails | null
    sizeRem: number
}


const baseStyle = css({
    backgroundColor: 'white',
    margin: '0.15rem'
})


export const Album: FC<Props> = ({ details, sizeRem }) => {
    const style = cx(
        baseStyle,
        css({
            width: `${sizeRem}rem`,
            height: `${sizeRem}rem`
        })
    )
    const image = details === null
        ? null
        : <img src={details.image.smallURL} alt={formatAlbumTitle(details)}/>
    return (
        <div className={style}>
            {image}
        </div>
    )
}
