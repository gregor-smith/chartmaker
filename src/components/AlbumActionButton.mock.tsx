const mock: typeof import('@/components/AlbumActionButton') = {
    AlbumActionButton: props =>
        <div {...props} className='mock-album-action-button'/>
}


export const AlbumActionButton = mock.AlbumActionButton
