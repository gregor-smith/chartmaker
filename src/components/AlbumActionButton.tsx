import { h, FunctionComponent } from 'preact'
import { css } from 'emotion'


type Props = {
    onClick: () => void
    title: string
}


const style = css({
    background: 'none',
    border: 'none'
})


export const AlbumActionButton: FunctionComponent<Props> = props =>
    <button {...props} class={style}/>
