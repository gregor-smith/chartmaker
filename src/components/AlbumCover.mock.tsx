import { cx } from 'emotion'


const mock: typeof import('@/components/AlbumCover') = {
    AlbumCover: ({
        album,
        size,
        overlayClass,
        children,
        highlighted,
        ...props
    }) => {
        const json = JSON.stringify(album)
        const className = cx('mock-album-cover', overlayClass)
        return (
            <div {...props} className={className}>
                {`Album: ${json}`}
                {`Size: ${size}`}
                {`Highlighted: ${highlighted}`}
                {children}
            </div>
        )
    }
}


export const AlbumCover = mock.AlbumCover
