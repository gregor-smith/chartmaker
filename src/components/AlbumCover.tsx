import React, { FC, DragEvent } from 'react'
import { css, cx } from 'emotion'

import { TEXT_COLOUR, ALBUM_PADDING_SIZE, highlightBackgroundStyle } from '@/style'
import { Album } from '@/types'


export type AlbumCoverProps = {
    album: Album
    size: string
    onDragStart: (event: DragEvent<HTMLDivElement>) => void
    onDragOver?: (event: DragEvent<HTMLDivElement>) => void
    onDragEnter?: (event: DragEvent<HTMLDivElement>) => void
    onDrop?: (event: DragEvent<HTMLDivElement>) => void
    onMouseEnter?: () => void
    overlayClass?: string
    highlighted?: boolean
}


const baseContainerStyle = css({
    position: 'relative',
    background: TEXT_COLOUR,
    margin: ALBUM_PADDING_SIZE
})

const imageStyle = css({
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    maxWidth: '100%',
    height: '100%',
    objectFit: 'contain'
})

// the overlay is given an additional non-emotion class so it can be used as a
// selector in the container's style
const overlaySelector = 'overlay'

const baseOverlayStyle = cx(
    overlaySelector,
    imageStyle,
    css({ zIndex: 10 })
)


export const AlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    onDragStart,
    onDragEnter,
    onDragOver,
    onDrop,
    onMouseEnter,
    children,
    overlayClass,
    highlighted
}) => {
    const overlayStyle = cx(baseOverlayStyle, overlayClass)
    const containerStyle = cx(
        baseContainerStyle,
        css({
            width: size,
            height: size,
            [`:not(:hover) .${overlaySelector}`]: {
                display: 'none'
            }
        }),
        highlighted === false
            ? highlightBackgroundStyle
            : undefined
    )

    let image: JSX.Element | undefined
    if (!album.placeholder) {
        image = (
            // lazy loading causes problems with html2canvas so keep it default
            <img className={imageStyle}
                src={album.url}
                alt={album.name}
                crossOrigin='anonymous'/>
        )
    }

    return (
        <div className={containerStyle}
                draggable={!album.placeholder}
                onDragStart={onDragStart}
                onDragEnter={onDragEnter}
                onDragOver={onDragOver}
                onDrop={onDrop}
                onMouseEnter={onMouseEnter}
                title={album.placeholder ? undefined : album.name}>
            {image}
            <div className={overlayStyle}>
                {children}
            </div>
        </div>
    )
}
