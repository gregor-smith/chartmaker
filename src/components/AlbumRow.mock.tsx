const mock: typeof import('@/components/AlbumRow') = {
    AlbumRow: ({ children }) =>
        <div className='mock-album-row'>
            {children}
        </div>
}


export const AlbumRow = mock.AlbumRow
