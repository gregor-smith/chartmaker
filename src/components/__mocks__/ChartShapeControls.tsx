import type { FC } from 'react'

import type { ChartShapeControlsProps } from '@/components/ChartShapeControls'


export const ChartShapeControls: FC<ChartShapeControlsProps> = ({
    shape,
    rowsX,
    rowsY
}) => {
    const json = JSON.stringify(shape)
    return (
        <div className='mock-chart-shape-controls'>
            {`Shape: ${json}`}
            {`Rows X: ${rowsX}`}
            {`Rows Y: ${rowsY}`}
        </div>
    )
}
