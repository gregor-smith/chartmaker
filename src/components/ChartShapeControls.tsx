import type { FC } from 'react'
import { css } from 'emotion'

import type { DispatchProps } from '@/reducer'
import type { ChartShape } from '@/types'
import { MAX_COLLAGE_ROWS_X, MAX_COLLAGE_ROWS_Y } from '@/constants'
import { ControlledSlider } from '@/components/ControlledSlider'
import { ControlledRadioButton } from '@/components/ControlledRadioButton'
import { SidebarGroup } from '@/components/SidebarGroup'


export type ChartShapeControlsProps = DispatchProps & {
    shape: ChartShape
    rowsX: number
    rowsY: number
}


const style = css({
    display: 'flex'
})


export const ChartShapeControls: FC<ChartShapeControlsProps> = ({
    dispatch,
    shape,
    rowsX,
    rowsY
}) => {
    function changeShape(shape: ChartShape) {
        dispatch({
            tag: 'UpdateChartShape',
            shape,
            rowsX,
            rowsY
        })
    }

    function switchToTop40() {
        changeShape({ tag: 'Top', size: 40 })
    }

    function switchToTop42() {
        changeShape({ tag: 'Top', size: 42 })
    }

    function switchToTop100() {
        changeShape({ tag: 'Top', size: 100 })
    }

    function switchToCollage() {
        changeShape({ tag: 'Collage' })
    }

    let rowsXSlider: JSX.Element | undefined
    let rowsYSlider: JSX.Element | undefined
    if (shape.tag === 'Collage') {
        const updateRowsX = (rowsX: number) =>
            dispatch({
                tag: 'UpdateChartShape',
                shape: { tag: 'Collage' },
                rowsX,
                rowsY
            })
        const updateRowsY = (rowsY: number) =>
            dispatch({
                tag: 'UpdateChartShape',
                shape: { tag: 'Collage' },
                rowsX,
                rowsY
            })

        rowsXSlider = (
            <ControlledSlider id='rows-x'
                    min={1}
                    max={MAX_COLLAGE_ROWS_X}
                    step={1}
                    value={rowsX}
                    onChange={updateRowsX}>
                X
            </ControlledSlider>
        )
        rowsYSlider = (
            <ControlledSlider id='rows-y'
                    min={1}
                    max={MAX_COLLAGE_ROWS_Y}
                    step={1}
                    value={rowsY}
                    onChange={updateRowsY}>
                Y
            </ControlledSlider>
        )
    }

    return (
        <SidebarGroup>
            <div className={style}>
                <ControlledRadioButton id='top40'
                        checked={shape.tag === 'Top' && shape.size === 40}
                        onCheck={switchToTop40}>
                    Top 40
                </ControlledRadioButton>
                <ControlledRadioButton id='top42'
                        checked={shape.tag === 'Top' && shape.size === 42}
                        onCheck={switchToTop42}>
                    Top 42
                </ControlledRadioButton>
                <ControlledRadioButton id='top100'
                        checked={shape.tag === 'Top' && shape.size === 100}
                        onCheck={switchToTop100}>
                    Top 100
                </ControlledRadioButton>
                <ControlledRadioButton id='collage'
                        checked={shape.tag === 'Collage'}
                        onCheck={switchToCollage}>
                    Collage
                </ControlledRadioButton>
            </div>
            {rowsXSlider}
            {rowsYSlider}
        </SidebarGroup>
    )
}
