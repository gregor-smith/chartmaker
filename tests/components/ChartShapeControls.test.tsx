import React from 'react'
import { render } from 'react-dom'
import { act } from 'react-dom/test-utils'

import {
    RenderContainer,
    ignore,
    fireEvent,
} from '../utils'
import { ChartShapeControls, ChartShapeControlsProps } from '@/components/ChartShapeControls'


jest.mock('@/components/SidebarGroup')
jest.mock('@/components/ControlledRadioButton')
jest.mock('@/components/ControlledSlider')


const container = new RenderContainer()


type ActionParams = Parameters<ChartShapeControlsProps['dispatch']>;


test.each([ 40, 42, 100 ] as const)('top shape renders radio buttons', size => {
    render(
        <ChartShapeControls dispatch={ignore}
            rowsX={10}
            rowsY={10}
            shape={{ tag: 'Top', size }}/>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})


test('collage shape renders radio buttons and sliders', () => {
    render(
        <ChartShapeControls dispatch={ignore}
            rowsX={10}
            rowsY={10}
            shape={{ tag: 'Collage' }}/>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})


test.each([ 40, 42, 100 ] as const)('checking top radio buttons dispatches action', size => {
    const mock = jest.fn<void, ActionParams>()

    render(
        <ChartShapeControls dispatch={mock}
            rowsX={5}
            rowsY={5}
            shape={{ tag: 'Collage' }}/>,
        container.element
    )

    act(() =>
        fireEvent(
            'change',
            container.element?.querySelector(`#top${size}`)
        )
    )

    expect(mock).toHaveBeenCalledTimes(1)
    expect(mock).toHaveBeenCalledWith<ActionParams>({
        tag: 'UpdateChartShape',
        shape: { tag: 'Top', size },
        rowsX: 5,
        rowsY: 5
    })
})


test('checking collage radio button dispatches action', () => {
    const mock = jest.fn<void, ActionParams>()

    render(
        <ChartShapeControls dispatch={mock}
            rowsX={5}
            rowsY={5}
            shape={{ tag: 'Top', size: 100 }}/>,
        container.element
    )

    act(() =>
        fireEvent(
            'change',
            container.element?.querySelector('#collage')
        )
    )

    expect(mock).toHaveBeenCalledTimes(1)
    expect(mock).toHaveBeenCalledWith<ActionParams>({
        tag: 'UpdateChartShape',
        shape: { tag: 'Collage' },
        rowsX: 5,
        rowsY: 5
    })
})


test('moving collage x slider dispatches action', () => {
    const mock = jest.fn<void, ActionParams>()

    render(
        <ChartShapeControls dispatch={mock}
            rowsX={5}
            rowsY={5}
            shape={{ tag: 'Collage' }}/>,
        container.element
    )

    act(() =>
        fireEvent(
            'change',
            container.element?.querySelector('#rows-x'),
            { target: { value: '6' } }
        )
    )

    expect(mock).toHaveBeenCalledTimes(1)
    expect(mock).toHaveBeenCalledWith<ActionParams>({
        tag: 'UpdateChartShape',
        shape: { tag: 'Collage' },
        rowsX: 6,
        rowsY: 5
    })
})


test('moving collage y slider dispatches action', () => {
    const mock = jest.fn<void, ActionParams>()

    render(
        <ChartShapeControls dispatch={mock}
            rowsX={5}
            rowsY={5}
            shape={{ tag: 'Collage' }}/>,
        container.element
    )

    act(() =>
        fireEvent(
            'change',
            container.element?.querySelector('#rows-y'),
            { target: { value: '4' } }
        )
    )

    expect(mock).toHaveBeenCalledTimes(1)
    expect(mock).toHaveBeenCalledWith<ActionParams>({
        tag: 'UpdateChartShape',
        shape: { tag: 'Collage' },
        rowsX: 5,
        rowsY: 4
    })
})
