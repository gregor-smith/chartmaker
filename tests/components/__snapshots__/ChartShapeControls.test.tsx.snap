// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collage shape renders radio buttons and sliders 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  id="test-render-container"
>
  <div
    class="mock-sidebar-group"
  >
    <div
      class="emotion-0"
    >
      <div
        class="mock-controlled-radio-button"
        id="top40"
      >
        Checked: false
        Top 40
      </div>
      <div
        class="mock-controlled-radio-button"
        id="top42"
      >
        Checked: false
        Top 42
      </div>
      <div
        class="mock-controlled-radio-button"
        id="top100"
      >
        Checked: false
        Top 100
      </div>
      <div
        class="mock-controlled-radio-button"
        id="collage"
      >
        Checked: true
        Collage
      </div>
    </div>
    <div
      class="mock-controlled-slider"
      id="rows-x"
    >
      Min: 1
      Max: 10
      Step: 1
      Value: 10
      Disabled: undefined
      X
    </div>
    <div
      class="mock-controlled-slider"
      id="rows-y"
    >
      Min: 1
      Max: 10
      Step: 1
      Value: 10
      Disabled: undefined
      Y
    </div>
  </div>
</div>
`;

exports[`top shape renders radio buttons 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  id="test-render-container"
>
  <div
    class="mock-sidebar-group"
  >
    <div
      class="emotion-0"
    >
      <div
        class="mock-controlled-radio-button"
        id="top40"
      >
        Checked: true
        Top 40
      </div>
      <div
        class="mock-controlled-radio-button"
        id="top42"
      >
        Checked: false
        Top 42
      </div>
      <div
        class="mock-controlled-radio-button"
        id="top100"
      >
        Checked: false
        Top 100
      </div>
      <div
        class="mock-controlled-radio-button"
        id="collage"
      >
        Checked: false
        Collage
      </div>
    </div>
  </div>
</div>
`;

exports[`top shape renders radio buttons 2`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  id="test-render-container"
>
  <div
    class="mock-sidebar-group"
  >
    <div
      class="emotion-0"
    >
      <div
        class="mock-controlled-radio-button"
        id="top40"
      >
        Checked: false
        Top 40
      </div>
      <div
        class="mock-controlled-radio-button"
        id="top42"
      >
        Checked: true
        Top 42
      </div>
      <div
        class="mock-controlled-radio-button"
        id="top100"
      >
        Checked: false
        Top 100
      </div>
      <div
        class="mock-controlled-radio-button"
        id="collage"
      >
        Checked: false
        Collage
      </div>
    </div>
  </div>
</div>
`;

exports[`top shape renders radio buttons 3`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  id="test-render-container"
>
  <div
    class="mock-sidebar-group"
  >
    <div
      class="emotion-0"
    >
      <div
        class="mock-controlled-radio-button"
        id="top40"
      >
        Checked: false
        Top 40
      </div>
      <div
        class="mock-controlled-radio-button"
        id="top42"
      >
        Checked: false
        Top 42
      </div>
      <div
        class="mock-controlled-radio-button"
        id="top100"
      >
        Checked: true
        Top 100
      </div>
      <div
        class="mock-controlled-radio-button"
        id="collage"
      >
        Checked: false
        Collage
      </div>
    </div>
  </div>
</div>
`;
